### DISCUSSION TITLE
Ground System and Remote Dev Environments

### DISCUSSION THREAD

**Discussion by ugo94490** [nasa/cFS#307] [Ideas]
Hi everyone,

Recently we had several ideas for improvements with @ThibFrgsGmz and we wanted to post them here to get some feedback from the community

# GDS (GroundSystem)

Currently, the [cFS-GroundSystem](https://github.com/nasa/cFS-GroundSystem) provided with the framework is only a development tool.

We think it would be good to first migrate the GDS to a browser-based application.

This would allow to have a development environment closer to a "remote development environment".

For example with F Prime it is possible to run the framework in a container with GitPod (see https://github.com/nasa/fprime/pull/788).

We thought of several others features that could be added to the GDS that would allow us to test our software created with cFS more easily.

For example the launch of validation script that would allow us to:
- to launch commands at specific time or time relative to previous command.
- define the expected telemetry value at a more or less precise range.
- generate test reports.

We have also thought it would be nice to be able to visualize the telemetry as a graph.
Either by creating this chart system from scratch or by using a tool like [YAMCS](https://yamcs.org/) / [OpenMCT](https://nasa.github.io/openmct/) ?

We have seen that there was already a software that includes the different test functionalities mentioned above ([COSMOS](https://cosmosc2.com/) implemented in [OpenSatKit](https://github.com/OpenSatKit/OpenSatKit)), unfortunately this software is not a web application and therefore does not fit into a "remote development environment" initiative.

This new initiative seems to us very important as it allows to fully integrate cFS in a DevOps circuit.

It will allow all teams to have a "code anywhere" and "ready to code" environment which is essential for the operational team for example which does not necessarily manipulate all these tools / libs / languages. 

They will only have to launch GitPod / CodeSpaces / Docker for example to test the software via this web interface.

<br/>
We would be delighted to hear your opinions on these various improvements and to discuss with you in order to improve them or to add others

**Comment by stanislaw** [comment]
Quick comment on this specific point:

> We have seen that there was already a software that includes the different test functionalities mentioned above (COSMOS implemented in OpenSatKit), unfortunately this software is not a web application and therefore does not fit into a "remote development environment" initiative.

Check out this: https://github.com/BallAerospace/COSMOS/discussions/1316.

**Reply by ugo94490** [reply]
Thank you very much I had not noticed this issue !

**Comment by astrogeco** [comment]
I love this idea and its something that I would like to have available for the cFS community. Unfortunately I have yet to see a web based ground system that can do both command and telemetry. I might be wrong but OpenMCT only seems to do the telemetry side, granted, I haven't spent much time with it. Once COSMOS 5 is out we'll have an option but I think having something "bare bones" like the current python ground system that runs natively on a web browser (maybe even serverless?) would be a huge advantage.

**Reply by ugo94490** [reply]
Actually, OpenMCT does not have a native command interface.
It only allows to visualize telemetry but there is a plugin [that allows to link it to YAMCS](https://github.com/nasa/openmct/discussions/3854) in order to send commands.

COSMOS 5 seems to be a good option, moreover it will allow to "delegate" the GDS part. The contributors of the COSMOS repo are very active so the cFS GDS would stay permanently updated via a submodule (would replace the current cFS-GroundSystem submodule) and would have more and more functionality.

We think that some improvements can be made to COSMOS such as the creation of validation scenarios described abstractly in XML (like for EDS) which would seem to be easier to use for operational/production people. 
Apart from that COSMOS looks really complete.

I think a temporary ground system even serverless would be a big advantage as it would allow to start creating a remote development environment now, and it will be updated later with the integration of COSMOS 5.

**Reply by ThibFrgsGmz** [reply]
A small addition regarding the validation scenarios / test cases described in XML that can be used by both developers and operational staff.

Another reason why we are thinking about XML is that an interesting implementation appears in this publication [A Testing Framework for Critical Space SW](https://www.researchgate.net/publication/290814468_A_Testing_Framework_for_Critical_Space_SW), on page 5.

We think that a production person would handle an XML markup language more easily than a programming language, but maybe we are wrong.

**Reply by astrogeco** [reply]
I'm partial to YAML because it's easy to read

**Reply by jphickey** [reply]
IMO neither XML nor YAML is well-suited to describing something like a test case.  It's a markup language (the "ML") not a sequence language.

The simple test cases can probably be described (where inputs and expected outputs are relatively fixed), but as soon as you get into test cases that require some type of conditional (if this, then that result), or if a result is relative (expected commandcounter = last commandcounter + 1) then a markup language falls apart.

Embeddable script languages like Lua offer a lot of flexibility in this regard.  It "reads" pretty easily, and you can also embed snippets of code into other non-code documents.  It's pretty simple in the basic use-cases, but its still a full blown scripting language when you need it.  I like Lua.

**Reply by mprather** [reply]
We reviewed OpenMCT last year to see if it could be leverage within an existing application. At the time, it would not work well with other apps (e.g. ability to only use a few select classes/features was no there) - it was expecting to effectively be the app itself. 

As for testing languages and YAML/XML... I haven't seen a good testing framework that solely relies on descriptive markup.  I would recommend staying with a native code framework. This makes it much easier for development teams while leveraging the full power of the native language while minimizing the number of tools and languages needed to get the system working.

**Comment by astrogeco** [comment]
@ugo94490 on a side note, can we rename this topic to something like "Ground System and Remote Dev Environments" since GDS isn't a common acronym?

**Reply by ugo94490** [reply]
It's done !

**Comment by dmccomas** [comment]
The NASA/Goddard apps have test scenarios and test scripts. I
would recommend using some of these as inputs to evaluate a potential
validation technology/platform.  I think you'll find, as Joe mentioned,
that many of the scenarios require conditional logic. In my testing
experience I found STOL, the scripting language used by Goddard ground
systems, to be limiting. I was on one mission that used Python as the test
scripting language and IMO our testing improved in terms of productivity
and in the quality of test scripts because we could leverage python
language features.

**Reply by ThibFrgsGmz** [reply]
Python is indeed an interesting choice. It is also the language used to describe validation/regression scenarios in the ESA TASTE framework (https://taste.tuxfamily.org/wiki/index.php?title=Technical_topic:_Advanced_testing_with_MSC_and_Python_scripts).

From your experience with COSMOS with [OpenSatKit](https://github.com/OpenSatKit/OpenSatKit), is Python better than Ruby for this kind of scripting?

**Reply by astrogeco** [reply]
From my limited experience, I think Python's large userbase makes it a strong choice.

**Reply by dmccomas** [reply]
From a language perspective I think they both bring more than enough horsepower for testing so it's hard to state one is better than the other. Culturally I agree that Python' large userbase especially in the flight software community makes it a good choice.  
I also think the test framework is very important.   I was prototyping some COSMOS 4.* extensions in OSK but never matured them. COSMOS 5.0 is a complete departure from 4.* and the long term support for COSMOS 4.* is unknown.  I'm in the process of refactoring OSK into multiple modules so COSMOS will not be part of the core module.

**Comment by avan989** [comment]
Is the idea to implement this as "Ground System As A Service", hosted on AWS for people to test? or is the idea to create an local web applications running on the local machine that acts as a ground system?

**Reply by ugo94490** [reply]
We were thinking more of a local web application, so that each person will be free to do what he wants with it (host it online or not, make changes, etc...)

**Reply by avan989** [reply]
I think it’s doable but it would be tough to find someone with enough time to implement the entire thing. Is the idea to get the cfs team at nasa to build it or something that the open source community would do?

**Reply by jphickey** [reply]
FWIW, much of the "local web service" concept is already implemented using EDS, if you want to go that route.  The actual Software Bus - CGI bridge app exists but unfortunately wasn't part of the set that went through release authorization, so its NASA-only right now...  but it is definitely doable and has been done.

**Reply by ThibFrgsGmz** [reply]
@avan989 We are not necessarily looking to assign this development task to anyone in particular but rather to see if others share our idea. If this is indeed the case, we could imagine that a detailed issue is created, and that a collaborator takes the action to initiate this work. We could imagine a partner company taking on an intern to initiate something for example.

**Reply by ThibFrgsGmz** [reply]
@jphickey The link between S-EDS and the "local wel service" concept is not so obvious to me, could you elaborate if you don't mind?
I understand that the S-EDS concept is more or less still at the Work In Progress stage on the CCSDS SOIS side and that feedback from the SAVOIR group is being done to improve the DoT. Once the discussions around the data descriptive format standard are completed, we could more easily interface a Ground System with a Space System.

What does the NASA-only application you mention actually do?

PS: what does the acronym CGI stand for, please? I have trouble seeing what it means.

**Comment by astrogeco** [comment]
I think the word "ground system" is throwing us all off. I wonder if there's a better word for what @ugo94490 is asking for, which, in my understanding, is a web-browser based tool to send commands and read telemetry during developers. I think the main advantage of this over our current Python tool is that it could be easy to containerize and portable.

**Reply by mprather** [reply]
I would first ask what is the goal?  The current python-based tools are ugly as sin but it's not a show stopper by any means. Are they mostly portable? Yeah, mostly. The dialogs can be used on Windows machines via a xserver channel and they clearly work on a native linux machine as is.  

The real problem, from my perspective, is how the config files are created/maintained. The "parse a header" approach is just hokey. We replaced that process with a simple script that makes it easy to adjust and completely removes the reliance on the header fitting an exact, yet fragile format.  A change such as that would be a would be a nice, simple win for the cFS team with no real additional infrastructure or tooling needed.

If you want portable, then I'm not sure advocating for web app that requires a container or new infrastructure is the right approach. To me that is just making the whole process way more difficult. Yes, it will be pretty but are you buying more functionality? Not really.  Are you buying increased complexity? Yes. Are you making it that much harder to support cFS (for both the product team and the consumers)? Probably. Will it be easier for newbies to pick up and extend cFS? Nope.

I guess in some way it's all about perspective -- I wouldn't use the cfs-GroundSystem tools anywhere but a dev environment or a demo/poc. That's how our team has only every seen them pitched (or more to the point, how they should be pitched).  If you really want to facilitate testing and make life easier, it's rather easy to write your own tools that send commands and/or receive telemetry. Those items will become part of the production release. I don't want to stand up yet-another-web-app-somewhere to get dev tools up and running.

**Reply by jbohren-hbr** [reply]
> @astrogeco I think the word "ground system" is throwing us all off.

Yeah, as we started developing with cFS what we really wanted was a means to rapidly send/receive commands/telemetry from both scripts and the command line. Access over a web interface wasn't a high priority, we just needed to interface the cFS systems with other heritage and under-development software systems.

> @mprather The real problem, from my perspective, is how the config files are created/maintained. The "parse a header" approach is just hokey.

@mprather Yeah we're currently doing this (hokey header parsing with [pyclibrary](https://pyclibrary.readthedocs.io/en/latest/index.html)) with a system we built in-house, but we generate ROS interface definition files from the headers to use with our other tooling. We'd like to leverage the EDS specification to generate both cFS message headers and ROS idl files, but there are still a lot of constants defined in cFS headers which find their way into the messages (such as `CFE_MISSION_MAX_PATH_LEN` which would end up being maintained in multiple places.

@astrogeco in https://github.com/nasa/cFS/discussions/284, @jphickey mentioned that some work has been done to use EdsLib to generate message headers. Is there an open-source generator that can use the EDS files as the *single source of truth* for message and configuration header generation?

**Reply by ThibFrgsGmz** [reply]
Thanks for sharing your opinion @mprather, it's interesting. With @ugo94490 we think it's a question of perspective because we don't quite agree with you on the Remote Development Environment.

## Introduction

We believe that we have a definition, perhaps more global, progressive or cutting-edge, of what a framework is, and what it should aim for through an iterative, incremental process.

It should cover as many phases of the flight software life cycle as possible, it should centralize the efforts of the community in our field as much as possible and centralize the costs worldwide in order to have the most versatile platform possible.
With this, we would have many more companies to energize the space sector to focus only on new applications, and make transparent the hardware issues, communication protocol standards or others.

Due to the convergence of space technologies and their standardization (e.g. LEON 3 CPU, Spacewire or Time-Triggered Ethernet Protocol), it is now more feasible and interesting to design this generic framework that can be used as an off-the-shelf product in future space projects.

We sense more or less a willingness from NASA to have private companies adopt their cFS framework because it is recommended on the Lunar Gateway and a CCSDS on cFS is being written to make its architecture a reference for space systems.

On the other hand, we see an emerging competition between the flight software frameworks whose number is growing (TASTE, GERICOS, F Prime, NanosatMO, KubOS,...).

Although the cFS architecture and its implementation are among the best, we see a lack of tools in the cFS bundle to cover more phases of the space project life cycle.

To gain strength and impose itself internationally we think it should integrate or be inspired by the efforts of each distributor (OpenSatKit, NOS3, Icarous, ...) to gather the community around a common project and to be able to propose/work on disruptive technologies to implement.

We will try to explain our point of view on what a framework should do by stating other more or less concrete needs.

## Feasibility study/evaluation aspect:

As we all know in space instrument projects the hardware is only available very late and is in a time lag with the flight software development.

The disadvantage is that it is necessary for the flight software team to be able to mock up the project as soon as the feasibility phase is over, i.e. well before the hardware design.

These mock-ups allow, among other things, to dimension the hardware platform by evaluating the CPU load, the amount of memory required or the upstream and downstream data rates.

So this raises a need for teams to have an environment that allows them to simulate/emulate hardware platforms for rapid prototyping, a sort of software testbed.

If the framework has such tools, it is easy to imagine teams instantiating a project container on a development server and working on it remotely from their laptop. In the same way as with GitPod, GitHub CodeSpace, Remote Repositories/Containers of VSCode.

By doing this the team would be directly operational as all installation phases subject to human error or portability/compatibility would be non-existent.

Once the project container is up and running the team would have a simulated/emulated platform flight software to study the performance of the system.
The environment would be ready to use, the GUI (Ground System) could display the curves of characteristic metrics easily exploitable for performance assessments, benchmarks and trade-offs.

If we go further we can imagine to have the possibility to switch from the simulated platform to the hardware platform or from simulated instruments to hardware instruments.

## Development cycle aspect

Tools allowing software engineers to automate repetitive tasks such as generating a code skeleton of a component/application and a corresponding unit test skeleton.

To have a tool adopting an MBSE approach (in AADL language ?) to model the topology of the system with a common environment between system team and flight software. An MBSE approach with self-generating skeleton code could make transparent an implementation of the system in C, C++, Ada or Rust language.

A pre-built application suite (AppStore) providing specific and common services to several projects that the developer can choose to integrate or not in his system.

A tool for running validation/regression scenarios with definition of command and telemetry messages taking a time scale and generating a results report.

## AIT/Production phase aspect

The AIT team in charge of integration testing needs an interactive graphical means to create test scenarios, display/save measurements, log commands sent and measurements received, or generate a complete test report.

## Conclusion

These are only ideas for improvement, they are obviously not a strict truth and only commit us. Some of these ideas are already implemented by competing frameworks or even cFS distributions.

In case we would like to integrate this kind of functionality quickly, it might be interesting to integrate some features of cFS distributions into the bundle as we go along.

From what we understand this is not the priority of the cFS program team or even their intention.

**Reply by jbohren-hbr** [reply]
@ThibFrgsGmz you've touched on a lot of topics above, and we share a lot of your concerns regarding development environment and a desire for more model-driven design. Would you agree that the first hurdle is to standardize specification and tooling around interface definitions such as SOIS EDS or equivalent?

**Reply by ThibFrgsGmz** [reply]
@jbohren-hbr Indeed, this is perhaps the highest priority barrier in that this type of standard provides an abstraction to detail the specification of a system's data types and the constraints associated with each. This allows a set of interactions between systems/components to be defined by thinking in terms of Service Data Units (SDUs) without having visibility into the implementation elements, including the programming language.

I think SOIS EDS is the best choice for two reasons that immediately come to my mind.
* The European [SAVOIR-SAFI]( https://savoir.estec.esa.int/SAVOIROrganisation.htm) group working on standardizing functional interfaces for AOCS sensors and actuators specifies that device manufacturers shall provide an EDS used to generate the sensor/actuator driver, which in turn is used by the software to connect the device to the bus (see here https://essr.esa.int/project/osra-onboard-software-reference-architecture accessible only to Europeans if I remember correctly).
* The [International Deep Space Standard]( https://www.internationaldeepspacestandards.com/) for Software, specifically ISwSIS, specifies the use of SOIS EDS for onboard interface services (see here https://nasasitebuilder.nasawestprime.com/idss2/wp-content/uploads/sites/45/2020/10/software_baseline_final_9-2020.pdf).

I don't really know why the adoption of SOIS EDS is so slow.
* Is there a lack of user-friendly tools in Python for example, easy to use and to integrate in a buildchain?
* Does the DoT not meet their needs because they don't cover all the mission-specific requirements?
* Do they prefer JSON or YAML to XML for reasons of readability for example?
* Is it a question of technological heritage and do companies prefer to reuse their proprietary tools?

**Reply by jbohren-hbr** [reply]
> I don't really know why the adoption of SOIS EDS is so slow.

I think right now a challenge is (a)  making the existing tooling accessible, and (b) standardizing an invocation method of that tooling for a cFS system.

Right now, even the EdsLib documentation reads:

> this tool is generally _not_ intended to be executed manually [[ref]](https://github.com/nasa/EdsLib/blob/master/doc/eds-dom-reference.md#invocation)

@jphickey's example integration of NASA's EdsLib with cFS [[ref]](https://github.com/jphickey/cfe-eds-framework) shows how EdsLib can be integrated with the cFS buildsystem. That example fully generates all of the platform and message-specific headers from EDS xml files.

Unfortunately, the implementation is so tightly coupled with the cFS sources that anyone who started using it would have to maintain the tooling and additional headers themselves.

Maybe a good first step is creating a modularized version of the EDS support using EdsLib so that people can start using it for their platform configuration headers and app headers. This would be something you could drop into a cfs mission definition folder without having to modify the core modules.

**Reply by astrogeco** [reply]
This thorough writeup and thoughtful conversation completely encapsulate the cFS community's potential. I'm grateful for every bit of knowledge and insight from you all. I don't have any steps forward right now but I'm definitely using this thread to advocate for more support for the community from an organizational perspective.

**Reply by ThibFrgsGmz** [reply]
@jbohren-hbr In my opinion, the source generation tool should be totally decoupled from the cFS framework otherwise it will be difficult to be adopted by other space actors.

The tool should take a suite of SEDS with a source code template, defined with [CookieCutter](https://cookiecutter.readthedocs.io/en/1.7.2/index.html) for example, to make it flexible and adaptable over time.

The developer would then be free to define the source code templates (C, C++, Ada, Rust...) that respect the coding/naming rules of his framework/company.

---

By the way, regarding the Remote Dev Env, we are completely in line with @astroesteban 's [demonstration](https://github.com/nasa/cFS/discussions/342#discussion-3529197) on cFS, it goes exactly in the same direction as we imagined/have sketched.

**Reply by jbohren-hbr** [reply]
> @jbohren-hbr In my opinion, the source generation tool should be totally decoupled from the cFS framework otherwise it will be difficult to be adopted by other space actors.

Yeah, ideally there's a clear separation managed by a well-defined interface in the build infrastructure, and EdsLib _is_ designed that way. I didn't mean to imply that EdsLib couldn't be used outside of cFS. However, the current instantiation of cfe which has been modified to leverage EdsLib is not backwards-compatible with the traditional non-EDS cFS API.

For example, EdsLib can (and should) be used to do more than just generate interface headers. It's also generating runtime sources and other facilities which address some common pitfalls around using cFS like message ID portability and target endainness management.

Currently, the build-time dependency graph for cfe when using EdsLib looks something like the following:
```

                               ┌──────────┐
                           ┌───┤eds2cfetbl│◄─────────────────────────────────────────────────────────────────┐
                           │   └──────────┘                                                                  │
                           │                                                                                 │
                           │   ┌──────────┐                                                                  │
                           │   │          │◄─────────────────────────────────────────────────────────────────┤
                           ├───┤missionlib│                                                                  │
                           │   │          │◄────────┐                                                        │
             ┌────────┐    │   └──────────┘         │                                                        │
             │ EdsLib │◄───┤                        │                                                        │   ┌───────┐
             └────────┘    │   ┌──────────┐         │                                                        ├───┤  cfe  │
                           └───┤ sedstool │◄────────┤                                                        │   └───────┘
                               └──────────┘         │ ┌───────────────────────┐    ┌────────────────────┐    │
                                                    ├─┤ cfe generated sources │◄───┤ cfe static sources │◄───┘
  ┌────────────────────┐      ┌───────────────┐     │ └───────────────────────┘    └────────────────────┘
  │ mission XML config │◄─────┤ cfe XML files │◄────┘
  └────────────────────┘      └───────────────┘
```

So the source generation tools (`EdsLib` and `sedstool`) are independent, but there is information in the mission EDS XML configuration which influences array sizes and other parameters in the cfe EDS XML files, which are managed within the cfe sources, and are finally used to generate headers and source files. Then there is a set of buildtime and runtime facilities like `missionlib` which are necessary to build and run cfe with EDS support. Finally, `eds2cfetbl` is needed by any cfe apps which generate tables.

In this way, even though EdsLib can be used outside of cfe, the current EDS-aware version of cfe can't easily be used without EdsLib. We're interested in helping address this challenge to make EDS utilization in cFS more accessible and less risky for adoption by the community.
